sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala

```diff
@@ -600,3 +604,146 @@ case class Logarithm(left: Expression, right: Expression)
     """
   }
 }
+
+case class Round(child: Expression, scale: Expression) extends Expression with ExpectsInputTypes {
+
+  def this(child: Expression) = {
+    this(child, Literal(0))
+  }
+
+  override def children: Seq[Expression] = Seq(child, scale)
+
+  override def nullable: Boolean = true
+
+  override def foldable: Boolean = child.foldable
+
+  override lazy val dataType: DataType = child.dataType match {
+      case DecimalType.Fixed(p, s) => DecimalType(p, _scale)
+      case t => t
+    }
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(
+    // rely on precedence to implicit cast String into Double
+    TypeCollection(DecimalType, DoubleType, FloatType, LongType, IntegerType, ShortType, ByteType),
+    TypeCollection(LongType, IntegerType, ShortType, ByteType))
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    child.dataType match {
+      case _: NumericType => // satisfy requirement
+      case dt =>
+        return TypeCheckFailure(s"Only numeric type is allowed for ROUND function, got $dt")
+    }
+    scale match {
+      case Literal(value, LongType) =>
+        if (value.asInstanceOf[Long] < Int.MinValue || value.asInstanceOf[Long] > Int.MaxValue) {
+          return TypeCheckFailure("ROUND scale argument out of allowed range")
+        }
+      case _ =>
+        if (scale.dataType.isInstanceOf[IntegralType] && scale.foldable) {
+          // TODO: How to check out of range for foldable LongType Expression
+          // satisfy requirement
+        } else {
+          return TypeCheckFailure("Only foldable Integral Expression " +
+            s"is allowed for ROUND scale arguments, got ${child.dataType}")
```

Comment:
If you accept only IntegerType for the 2nd argument, long type will get implicitly casted to integer by the optimizer. so I think you can just check whether it is foldable.

also you should call super.checkInputDataTypes() to make sure the generic type checker can run successfully, and only run the 2nd argument foldable check after that.